### PR TITLE
Byte buf tls ctx

### PR DESCRIPTION
**Pull request by JonathanHenson**
Tls initialization now uses byte buffers, byte cursors etc... instead of file paths. As a result, we have new ownership semantics, copies etc....

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/io/tls_channel_handler.h [hunk 1]
```diff
@@ -88,6 +90,8 @@ struct aws_tls_connection_options {
 };
 
 struct aws_tls_ctx_options {
+    struct aws_allocator *allocator;
```
Comment (justinboswell) [inline include/aws/io/tls_channel_handler.h]: If you don't want people manipulating this manually, is it worth putting it in a private header and just forward declaring it for pointers here?

File: source/darwin/secure_transport_tls_channel_handler.c [hunk 2]
```diff
@@ -922,72 +925,42 @@ static struct aws_tls_ctx *s_tls_ctx_new(struct aws_allocator *alloc, struct aws
...
-        struct aws_byte_buf password = {.buffer = NULL, .len = 0, .allocator = NULL, .capacity = 0};
-
-        if (options->pkcs12_password) {
-            password = aws_byte_buf_from_c_str(options->pkcs12_password);
-        }
-
-        struct aws_byte_cursor pkcs12_blob_cur = aws_byte_cursor_from_buf(&pkcs12_blob);
-        struct aws_byte_cursor password_cur = aws_byte_cursor_from_buf(&password);
+    } else if (options->pkcs12.len) {
+        struct aws_byte_cursor pkcs12_blob_cur = aws_byte_cursor_from_buf(&options->pkcs12);
```
Comment (justinboswell) [inline source/darwin/secure_transport_tls_channel_handler.c]: Debug log message?

File: source/tls_channel_handler.c [hunk 3]
```diff
@@ -14,77 +14,266 @@
...
+    if (options->ca_path) {
+        aws_string_destroy(options->ca_path);
+    }
+
+    if (options->certificate.len) {
+        aws_byte_buf_clean_up(&options->certificate);
+    }
+
+    if (options->private_key.len) {
+        aws_secure_zero(options->private_key.buffer, options->private_key.len);
```
Comment (justinboswell) [inline source/tls_channel_handler.c]: aws_byte_buf_clean_up_secure

File: source/tls_channel_handler.c [hunk 4]
```diff
@@ -14,77 +14,266 @@
...
+        aws_byte_buf_clean_up(&options->private_key);
+    }
+
+#ifdef __APPLE__
+    if (options->pkcs12.len) {
+        aws_secure_zero(options->pkcs12.buffer, options->pkcs12.len);
```
Comment (justinboswell) [inline source/tls_channel_handler.c]: aws_byte_buf_clean_up_secure

File: source/tls_channel_handler.c [hunk 5]
```diff
@@ -14,77 +14,266 @@
...
+        aws_byte_buf_clean_up(&options->pkcs12);
+    }
+
+    if (options->pkcs12_password.len) {
+        aws_secure_zero(options->pkcs12_password.buffer, options->pkcs12_password.len);
```
Comment (justinboswell) [inline source/tls_channel_handler.c]: aws_byte_buf_clean_up_secure

File: include/aws/io/tls_channel_handler.h (hunk 1 above)
Comment (JonathanHenson) [reply include/aws/io/tls_channel_handler.h]: I don't, but I also don't want yet another allocation failure point, branch nonsense in the user's code. Making it private means I have to have some return a pointer to something function, which I don't really want to do.

File: source/darwin/secure_transport_tls_channel_handler.c (hunk 2 above)
Comment (JonathanHenson) [reply source/darwin/secure_transport_tls_channel_handler.c]: the logging is now in the file_utils. I don't know what i'd log here.

File: source/tls_channel_handler.c (hunk 3 above)
Comment (JonathanHenson) [reply source/tls_channel_handler.c]: will do.

File: source/tls_channel_handler.c (hunk 4 above)
Comment (JonathanHenson) [reply source/tls_channel_handler.c]: will do

File: source/tls_channel_handler.c (hunk 5 above)
Comment (JonathanHenson) [reply source/tls_channel_handler.c]: will do

File: source/darwin/secure_transport_tls_channel_handler.c (hunk 2 above)
Comment (justinboswell) [reply source/darwin/secure_transport_tls_channel_handler.c]: For all the other branches, you AWS_LOGF_DEBUG that you're initializing

File: include/aws/io/tls_channel_handler.h (hunk 1 above)
Comment (justinboswell) [reply include/aws/io/tls_channel_handler.h]: That's fair.

File: source/darwin/secure_transport_tls_channel_handler.c (hunk 2 above)
Comment (JonathanHenson) [reply source/darwin/secure_transport_tls_channel_handler.c]: oh, damn github diffs, I thought this was a different c file. will do

**Pull request merged**